projects/ssr-tests/src/ssr-testing.spec.ts

```diff
@@ -1,54 +1,126 @@
-import * as Log from './log.utils';
-import * as ProxyServer from './proxy.utils';
-import * as Ssr from './ssr.utils';
+import { Server } from 'http';
+import { expectLogMessages } from './matchers/matchers';
+import * as LogUtils from './utils/log.utils';
+import * as ProxyUtils from './utils/proxy.utils';
+import * as SsrUtils from './utils/ssr.utils';
 
 const BACKEND_BASE_URL: string = process.env.CX_BASE_URL || '';
 
 describe('SSR E2E', () => {
-  let proxy: any;
+  let proxy: Server;
   const REQUEST_PATH = '/electronics-spa/en/USD/';
 
-  beforeEach(async () => {
-    Log.clearSsrLogFile();
-    await Ssr.startSsrServer();
+  beforeEach(() => {
+    LogUtils.clearSsrLogFile();
   });
 
   afterEach(async () => {
-    await proxy.close();
-    await Ssr.killSsrServer();
+    proxy.close();
+    await SsrUtils.killSsrServer();
   });
 
-  it('should receive success response with request', async () => {
-    proxy = await ProxyServer.startProxyServer({
-      target: BACKEND_BASE_URL,
-    });
-    const response: any = await ProxyServer.sendRequest(REQUEST_PATH);
-    expect(response.statusCode).toEqual(200);
-
-    // Rendering should not complete in the first request.
-    // App should fall back to csr.
-    Log.assertMessages([
-      `Rendering started (${REQUEST_PATH})`,
-      `Request is waiting for the SSR rendering to complete (${REQUEST_PATH})`,
-    ]);
-  });
+  describe('With SSR error handling', () => {
+    describe('Common behavior', () => {
+      beforeEach(async () => {
+        await SsrUtils.startSsrServer();
+      });
+
+      it('should receive success response with request', async () => {
+        proxy = await ProxyUtils.startBackendProxyServer({
+          target: BACKEND_BASE_URL,
+        });
+        const response: any =
+          await ProxyUtils.sendRequestToSsrServer(REQUEST_PATH);
+        expect(response.statusCode).toEqual(200);
+
+        // Rendering should not complete in the first request.
```

Comment:
[QUESTION]
Why? Because our `timeout` in `SsrOptimizationOptions` is only 3000s?
If yes, shouldn't we just increase the default `timeout` for e2e tests?

Reply:
That's a comment left by Caine, but it's actually a good question. I haven't noticed any issues with this by far so maybe the comment is not relevant anymore.

Reply:
Good point. In that case, I'm going to add another param to `startSsrServer` (with a default value of 10s? 20s?) so we can control the timeout.